### ISSUE TITLE
Support the ability to re-run build scripts if and only if the crate is being compiled for other reasons

### ISSUE THREAD

**Issue by quodlibetor** [rust-lang/cargo#8084]
**Describe the problem you are trying to solve**

We embed a bit of build-time metadata (e.g. git commit hash, wall-clock build time) inside our binaries. We specifically only would like to change this metadata on re-build, we don't want to force a rebuild because the metadata has changed. For example, checking out a commit that has only changed external documentation shouldn't retrigger a build, none of the source has changed. Also, when running tests, modifying the tests shouldn't necessarily retrigger a build of the library that the tests are running against.

**Describe the solution you'd like**
Perhaps a `cargo:rerun-on-compile=true` build.rs directive?

**Comment by ehuss** [comment]
I'm not sure I completely understand what you're asking for.

Would it be possible to use proc-macros instead of build scripts?  I'm not sure how you're embedding your metadata, but if you can include it as a static, that might work.

If you really need a build script, then I think if you emit `rerun-if-changed` for every source file for the binary (and/or library), it will essentially re-run every time the binary is built.  Maybe toss in `Cargo.toml` for good measure.  Is there some situation where that isn't good enough?

**Comment by quodlibetor** [comment]
> Would it be possible to use proc-macros instead of build scripts?

Yes, we're currently embedding statics using [compile-time-run](https://github.com/de-vri-es/rust-compile-time-run), and it's working fine, it's just could be an unnecessary dependency that makes our build process (potentially?) more fragile.

> emit rerun-if-changed for every source file .. Cargo.toml [and Cargo.lock]

Unfortunately that is hard to do when we've got vendored C dependencies.

A lot of [what we're trying to optimize for](https://github.com/MaterializeInc/materialize/commit/8ec395dfb9c529a05421c91085c6af1858d4268f) is making our integration tests not require a re-build of the library while still guaranteeing that we never get incorrect metadata in statics.
